`Cannot serialize property` error upon importing projects

With 2019.3 EAP, import any pants project will get the following error, which would annoy user.

```
com.intellij.serialization.SerializationException: Cannot serialize property (property=FieldAccessor(name=externalProjectStructure, class=com.intellij.openapi.externalSystem.model.internal.InternalExternalProjectInfo), binding=BeanBinding(beanClass=class com.intellij.openapi.externalSystem.model.DataNode), beanClass=com.intellij.openapi.externalSystem.model.internal.InternalExternalProjectInfo)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:82)
	at com.intellij.serialization.BaseCollectionBinding$createItemConsumer$1.accept(BaseCollectionBinding.kt:18)
	at com.intellij.concurrency.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4572)
	at com.intellij.serialization.CollectionBinding.serialize(CollectionBinding.kt:61)
	at com.intellij.serialization.IonObjectSerializer.doWrite(IonObjectSerializer.kt:119)
	at com.intellij.serialization.IonObjectSerializer.writeVersioned(IonObjectSerializer.kt:41)
	at com.intellij.serialization.VersionedFile.writeList(VersionedFile.kt:34)
	at com.intellij.openapi.externalSystem.service.project.manage.ExternalProjectsDataStorage.doSave(ExternalProjectsDataStorage.java:349)
	at com.intellij.openapi.externalSystem.service.project.manage.ExternalProjectsDataStorage.doSave(ExternalProjectsDataStorage.java:162)
	at com.intellij.configurationStore.SettingsSavingComponentJavaAdapter.save(SettingsSavingComponent.kt:16)
	at com.intellij.configurationStore.ComponentStoreWithExtraComponents$saveSettingsSavingComponentsAndCommitComponents$2$2.invokeSuspend(ComponentStoreWithExtraComponents.kt:71)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.kt:116)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:80)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at com.intellij.configurationStore.SaveAndSyncHandlerImpl.processTasks(SaveAndSyncHandlerImpl.kt:79)
	at com.intellij.configurationStore.SaveAndSyncHandlerImpl.access$processTasks(SaveAndSyncHandlerImpl.kt:47)
	at com.intellij.configurationStore.SaveAndSyncHandlerImpl$saveAlarm$1.invoke(SaveAndSyncHandlerImpl.kt:59)
	at com.intellij.configurationStore.SaveAndSyncHandlerImpl$saveAlarm$1.invoke(SaveAndSyncHandlerImpl.kt:47)
	at com.intellij.util.SingleAlarmKt$sam$java_lang_Runnable$0.run(SingleAlarm.kt)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:231)
	at com.intellij.util.Alarm$Request.runSafely(Alarm.java:371)
	at com.intellij.util.Alarm$Request.run(Alarm.java:358)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:217)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:222)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:30)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:201)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:221)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:190)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.intellij.serialization.SerializationException: Cannot serialize property (property=FieldAccessor(name=children, class=com.intellij.openapi.externalSystem.model.DataNode), binding=com.intellij.serialization.CollectionBinding@8c05533, beanClass=com.intellij.openapi.externalSystem.model.DataNode)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:82)
	at com.intellij.serialization.Binding$DefaultImpls.serialize(Binding.kt:10)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:19)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:79)
	... 37 more
Caused by: com.intellij.serialization.SerializationException: Cannot serialize property (property=FieldAccessor(name=data, class=com.intellij.openapi.externalSystem.model.DataNode), binding=com.intellij.serialization.PolymorphicBinding@45044679, beanClass=com.intellij.openapi.externalSystem.model.DataNode)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:82)
	at com.intellij.serialization.BaseCollectionBinding$createItemConsumer$1.accept(BaseCollectionBinding.kt:18)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at com.intellij.serialization.CollectionBinding.serialize(CollectionBinding.kt:61)
	at com.intellij.serialization.Binding$DefaultImpls.serialize(Binding.kt:10)
	at com.intellij.serialization.BaseCollectionBinding.serialize(BaseCollectionBinding.kt:8)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:79)
	... 40 more
Caused by: com.intellij.serialization.SerializationException: Cannot serialize property (property=FieldAccessor(name=myRoots, class=com.intellij.openapi.projectRoots.impl.ProjectJdkImpl), binding=BeanBinding(beanClass=class com.intellij.openapi.projectRoots.impl.RootsAsVirtualFilePointers), beanClass=com.intellij.openapi.projectRoots.impl.ProjectJdkImpl)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:82)
	at com.intellij.serialization.PolymorphicBinding.serialize(PolymorphicBinding.kt:14)
	at com.intellij.serialization.Binding$DefaultImpls.serialize(Binding.kt:10)
	at com.intellij.serialization.PolymorphicBinding.serialize(PolymorphicBinding.kt:4)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:79)
	... 46 more
Caused by: com.intellij.serialization.SerializationException: Cannot serialize property (property=FieldAccessor(name=myRoots, class=com.intellij.openapi.projectRoots.impl.RootsAsVirtualFilePointers), binding=com.intellij.serialization.MapBinding@22a926f6, beanClass=com.intellij.openapi.projectRoots.impl.RootsAsVirtualFilePointers)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:82)
	at com.intellij.serialization.Binding$DefaultImpls.serialize(Binding.kt:10)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:19)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:79)
	... 50 more
Caused by: java.lang.UnsupportedOperationException: byte is not supported
	at com.intellij.serialization.PrimitiveBindingsKt$registerPrimitiveBindings$byte$1.invoke(primitiveBindings.kt:34)
	at com.intellij.serialization.PrimitiveBindingsKt$registerPrimitiveBindings$byte$1.invoke(primitiveBindings.kt)
	at com.intellij.serialization.IonBindingProducer.getNestedBinding(IonBindingProducer.kt:103)
	at com.intellij.serialization.BeanBinding.init(BeanBinding.kt:37)
	at com.intellij.serialization.BindingProducer.getRootBinding(BindingProducer.kt:71)
	at com.intellij.serialization.IonBindingProducerKt.createElementBindingByType(IonBindingProducer.kt:184)
	at com.intellij.serialization.BaseCollectionBinding.<init>(BaseCollectionBinding.kt:9)
	at com.intellij.serialization.ArrayBinding.<init>(ArrayBinding.kt:7)
	at com.intellij.serialization.IonBindingProducer.createRootBinding(IonBindingProducer.kt:63)
	at com.intellij.serialization.BindingProducer.getRootBinding(BindingProducer.kt:68)
	at com.intellij.serialization.IonBindingProducer.getNestedBinding(IonBindingProducer.kt:111)
	at com.intellij.serialization.BeanBinding.init(BeanBinding.kt:37)
	at com.intellij.serialization.BindingProducer.getRootBinding(BindingProducer.kt:71)
	at com.intellij.serialization.IonBindingProducer.getNestedBinding(IonBindingProducer.kt:123)
	at com.intellij.serialization.BeanBinding.init(BeanBinding.kt:37)
	at com.intellij.serialization.BindingProducer.getRootBinding(BindingProducer.kt:71)
	at com.intellij.serialization.BindingProducer.getRootBinding(BindingProducer.kt:46)
	at com.intellij.serialization.PolymorphicBinding.serialize(PolymorphicBinding.kt:14)
	at com.intellij.serialization.MapBinding$serialize$1.invoke(MapBinding.kt:47)
	at com.intellij.serialization.MapBinding$serialize$3.execute(MapBinding.kt:70)
	at gnu.trove.THashMap.forEachEntry(THashMap.java:278)
	at com.intellij.serialization.MapBinding.serialize(MapBinding.kt:69)
	at com.intellij.serialization.Binding$DefaultImpls.serialize(Binding.kt:10)
	at com.intellij.serialization.MapBinding.serialize(MapBinding.kt:10)
	at com.intellij.serialization.BeanBinding.serialize(BeanBinding.kt:79)
	... 53 more
```

--- a/common/com/twitter/intellij/pants/model/JdkRef.java
+++ b/common/com/twitter/intellij/pants/model/JdkRef.java
@@ -1,0 +1,55 @@
+// Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+package com.twitter.intellij.pants.model;
+
+import com.intellij.openapi.projectRoots.JavaSdk;
+import com.intellij.openapi.projectRoots.ProjectJdkTable;
+import com.intellij.openapi.projectRoots.Sdk;
+import com.intellij.serialization.PropertyMapping;
+import com.twitter.intellij.pants.util.PantsUtil;
+
+import java.util.Objects;
+
+public final class JdkRef {
+  private final String myName;
+  private final String myHome;
+
+  public static JdkRef fromSdk(Sdk sdk) {
+    return new JdkRef(sdk.getName(), sdk.getHomePath());
+  }
+
+  @PropertyMapping({"myName", "myHome"})
+  private JdkRef(String myName, String myHome) {
+    this.myName = myName;
+    this.myHome = myHome;
+
+}
+  public Sdk toSdk() {
+    JavaSdk javaSdk = JavaSdk.getInstance();
+    return ProjectJdkTable.getInstance().getSdksOfType(javaSdk).stream()
+      .filter(sdk -> myHome.equals(sdk.getHomePath()))
+      .filter(sdk -> myName.equals(sdk.getName()))
+      .findFirst()
+      .orElseGet(() -> PantsUtil.createJdk(myName, myHome, null));
+  }
+
+  @Override
+  public String toString() {
+    return myName + " @ " + myHome;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (!(o instanceof JdkRef)) return false;
+    JdkRef ref = (JdkRef) o;
+    return Objects.equals(myName, ref.myName) &&
+           Objects.equals(myHome, ref.myHome);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(myName, myHome);
+  }
+}

--- a/common/com/twitter/intellij/pants/util/PantsConstants.java
+++ b/common/com/twitter/intellij/pants/util/PantsConstants.java
@@ -6,7 +6,7 @@ package com.twitter.intellij.pants.util;
 import com.intellij.openapi.externalSystem.model.Key;
 import com.intellij.openapi.externalSystem.model.ProjectKeys;
 import com.intellij.openapi.externalSystem.model.ProjectSystemId;
-import com.intellij.openapi.projectRoots.Sdk;
+import com.twitter.intellij.pants.model.JdkRef;
 import org.jetbrains.annotations.NotNull;
 
 
@@ -41,7 +41,7 @@ public class PantsConstants {
   public static final String PANTS_OPTION_ASYNC_CLEAN_ALL = "clean-all.async";
 
   // Used to initialize project sdk therefore use project processing weight, i.e, the highest.
-  public static final Key<Sdk> SDK_KEY = Key.create(Sdk.class, ProjectKeys.PROJECT.getProcessingWeight());
+  public static final Key<JdkRef> SDK_KEY = Key.create(JdkRef.class, ProjectKeys.PROJECT.getProcessingWeight());
 
   public static final String PANTS_CLI_OPTION_EXPORT_OUTPUT_FILE = "--export-output-file";
   public static final String PANTS_CLI_OPTION_LIST_OUTPUT_FILE = "--list-output-file";

--- a/common/com/twitter/intellij/pants/util/PantsUtil.java
+++ b/common/com/twitter/intellij/pants/util/PantsUtil.java
@@ -879,17 +879,24 @@ public class PantsUtil {
 
     // Finally if we need to create a new JDK, it needs to be registered in the `ProjectJdkTable` on the IDE level
     // before it can be used.
-    Sdk jdk = JavaSdk.getInstance().createJdk(jdkName, jdkHome.get());
-    ApplicationManager.getApplication().invokeAndWait(() -> {
-        ApplicationManager.getApplication().runWriteAction(() -> {
-            if (parentDisposable == null) {
-              ProjectJdkTable.getInstance().addJdk(jdk);
-            } else {
-              ProjectJdkTable.getInstance().addJdk(jdk, parentDisposable);
-            }
-        });
-    });
+    Sdk jdk = createJdk(jdkName, jdkHome.get(), parentDisposable);
     return Optional.of(jdk);
+  }
+
+  public static Sdk createJdk(String name, String home, Disposable disposable) {
+    Sdk jdk = JavaSdk.getInstance().createJdk(name, home);
+    ApplicationManager.getApplication().invokeAndWait(() -> {
+      ApplicationManager.getApplication().runWriteAction(() -> {
+        if (disposable == null) {
+          ProjectJdkTable.getInstance().addJdk(jdk);
+        }
+        else {
+          ProjectJdkTable.getInstance().addJdk(jdk, disposable);
+        }
+      });
+    });
+
+    return jdk;
   }
 
   /**

--- a/src/com/twitter/intellij/pants/service/project/PantsSystemProjectResolver.java
+++ b/src/com/twitter/intellij/pants/service/project/PantsSystemProjectResolver.java
@@ -40,6 +40,7 @@ import com.intellij.openapi.wm.ToolWindowManager;
 import com.intellij.util.Consumer;
 import com.intellij.util.messages.MessageBusConnection;
 import com.twitter.intellij.pants.metrics.PantsExternalMetricsListenerManager;
+import com.twitter.intellij.pants.model.JdkRef;
 import com.twitter.intellij.pants.projectview.PantsProjectPaneSelectInTarget;
 import com.twitter.intellij.pants.projectview.ProjectFilesViewPane;
 import com.twitter.intellij.pants.service.PantsCompileOptionsExecutor;
@@ -165,6 +166,7 @@ public class PantsSystemProjectResolver implements ExternalSystemProjectResolver
 
     PantsUtil.findPantsExecutable(executor.getProjectPath())
       .flatMap(file -> PantsUtil.getDefaultJavaSdk(file.getPath(), null))
+      .map(JdkRef::fromSdk)
       .ifPresent(sdk -> projectDataNode.createChild(PantsConstants.SDK_KEY, sdk));
 
     if (!isPreviewMode) {

--- a/src/com/twitter/intellij/pants/service/project/wizard/PantsProjectImportBuilder.java
+++ b/src/com/twitter/intellij/pants/service/project/wizard/PantsProjectImportBuilder.java
@@ -4,17 +4,15 @@
 package com.twitter.intellij.pants.service.project.wizard;
 
 import com.intellij.ide.util.projectWizard.WizardContext;
-import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.externalSystem.model.DataNode;
 import com.intellij.openapi.externalSystem.model.project.ProjectData;
 import com.intellij.openapi.externalSystem.service.project.ProjectDataManager;
 import com.intellij.openapi.externalSystem.service.project.wizard.AbstractExternalProjectImportBuilder;
 import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil;
 import com.intellij.openapi.project.Project;
-import com.intellij.openapi.projectRoots.JavaSdk;
-import com.intellij.openapi.projectRoots.ProjectJdkTable;
 import com.intellij.openapi.projectRoots.Sdk;
 import com.twitter.intellij.pants.PantsBundle;
+import com.twitter.intellij.pants.model.JdkRef;
 import com.twitter.intellij.pants.settings.ImportFromPantsControl;
 import com.twitter.intellij.pants.util.PantsConstants;
 import icons.PantsIcons;
@@ -22,7 +20,6 @@ import org.jetbrains.annotations.NotNull;
 
 import javax.swing.Icon;
 import java.io.File;
-import java.util.List;
 
 public class PantsProjectImportBuilder extends AbstractExternalProjectImportBuilder<ImportFromPantsControl> {
 
@@ -65,31 +62,10 @@ public class PantsProjectImportBuilder extends AbstractExternalProjectImportBuil
       return;
     }
 
-    final DataNode<Sdk> sdkNode = ExternalSystemApiUtil.find(node, PantsConstants.SDK_KEY);
+    final DataNode<JdkRef> sdkNode = ExternalSystemApiUtil.find(node, PantsConstants.SDK_KEY);
     if (sdkNode != null) {
-      Sdk pantsSdk = sdkNode.getData();
-      context.setProjectJdk(addIfNotExists(pantsSdk));
+      Sdk pantsSdk = sdkNode.getData().toSdk();
+      context.setProjectJdk(pantsSdk);
     }
-  }
-
-  /**
-   * Find if pants sdk is already configured, return the existing sdk if it exists,
-   * otherwise add to the config and return.
-   */
-  private Sdk addIfNotExists(final Sdk pantsSdk) {
-    final JavaSdk javaSdk = JavaSdk.getInstance();
-    List<Sdk> sdks = ProjectJdkTable.getInstance().getSdksOfType(javaSdk);
-    for (Sdk sdk : sdks) {
-      if (javaSdk.getVersion(sdk) == javaSdk.getVersion(pantsSdk)) {
-        return sdk;
-      }
-    }
-    ApplicationManager.getApplication().runWriteAction(new Runnable() {
-      @Override
-      public void run() {
-        ProjectJdkTable.getInstance().addJdk(pantsSdk);
-      }
-    });
-    return pantsSdk;
   }
 }

